extmod/modopenamp.c

```diff
@@ -0,0 +1,394 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * OpenAMP's Python module.
+ */
+#if MICROPY_PY_OPENAMP
+
+#include "py/obj.h"
+#include "py/nlr.h"
+#include "py/runtime.h"
+
+#include "metal/alloc.h"
+#include "metal/errno.h"
+#include "metal/io.h"
+#include "metal/sys.h"
+#include "metal/device.h"
+#include "metal/utilities.h"
+
+#include "openamp/open_amp.h"
+#include "openamp/remoteproc.h"
+#include "openamp/remoteproc_loader.h"
+#include "modopenamp.h"
+
+#if !MICROPY_ENABLE_FINALISER
+#error "MICROPY_PY_OPENAMP requires MICROPY_ENABLE_FINALISER"
+#endif
+
+#if MICROPY_PY_OPENAMP_RSC_TABLE_ENABLE
+#define VIRTIO_DEV_ID           0xFF
+#define VIRTIO_DEV_FEATURES     (1 << VIRTIO_RPMSG_F_NS)
+
+#define VRING0_ID               0   // VRING0 ID (host to remote) fixed to 0 for linux compatibility
+#define VRING1_ID               1   // VRING1 ID (remote to host) fixed to 1 for linux compatibility
+#define VRING_NOTIFY_ID         VRING0_ID
+
+#define VRING_COUNT             2
+#define VRING_ALIGNMENT         32
+// Note the number of buffers must be a power of 2
+#define VRING_NUM_BUFFS         64
+
+// The following config should be enough for about 128 descriptors.
+// See lib/include/openamp/virtio_ring.h for the layout of vrings
+// and vring_size() to calculate the vring size.
+#define VRING_RX_ADDR           (METAL_SHM_ADDR)
+#define VRING_TX_ADDR           (METAL_SHM_ADDR + 0x1000)
+#define VRING_BUFF_ADDR         (METAL_SHM_ADDR + 0x2000)
+#define VRING_BUFF_SIZE         (METAL_SHM_SIZE - 0x2000)
+
+static const char openamp_trace_buf[128];
+#define MICROPY_PY_OPENAMP_TRACE_BUF       ((uint32_t)openamp_trace_buf)
+#define MICROPY_PY_OPENAMP_TRACE_BUF_LEN   sizeof(MICROPY_PY_OPENAMP_TRACE_BUF)
+
+#endif // MICROPY_PY_OPENAMP_RSC_TABLE_ENABLE
+
+#define debug_printf(...)       // mp_printf(&mp_plat_print, __VA_ARGS__)
+
+#if MICROPY_PY_OPENAMP_RPROC
+extern mp_obj_type_t rproc_type;
+#endif
+
+static struct metal_device shm_device = {
```

Comment:
This (and in other places in this file) should be `STATIC` but we've decided it's finally time to get rid of that (see #13763) so actually before this gets merged we'll need to replace all the `STATIC` with `static` in this PR. (Sorry about that! #13763 has a script you can run)